### PR TITLE
Updates for GNSS API assistance usage

### PR DESCRIPTION
**Pull request by tokangas**
This PR updates GNSS API usage for multi-constellation assistance support.

Currently, the functionality remains as it was, i.e. only GPS assistance is used. QZSS assistance support with nRF Cloud will still be added into this PR.

### CODE REVIEW

**Comment by tokangas** [comment]
@lemrey This PR touches many components, but I'm not able to add reviewers. Would be nice to get reviews from some code owners as well.

File: lib/location/method_gnss.c [hunk 1]
```diff
@@ -381,12 +384,21 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
...
 #if defined(CONFIG_NRF_CLOUD_AGPS)
-bool method_gnss_agps_required(struct nrf_modem_gnss_agps_data_frame *request)
+bool method_gnss_agps_required(struct nrf_modem_gnss_agnss_data_frame *request)
 {
 	int32_t time_since_agps_req;
+	bool ephe_or_alm_required = false;
 
 	/* Check if A-GPS data is needed. */
-	if (request->sv_mask_ephe == 0 && request->sv_mask_alm == 0 && request->data_flags == 0) {
+	for (int i = 0; i < request->system_count; i++) {
```
Comment (tokangas) [inline lib/location/method_gnss.c:393]: This should only check GPS ephe/alm until QZSS assistance support is added.

**Comment by lemrey** [comment]
@tokangas could you try to add reviewers now?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c [hunk 2]
```diff
@@ -124,47 +124,60 @@ static void print_location(struct nrf_modem_gnss_agps_data_location *data)
...
+			(struct nrf_modem_gnss_agnss_gps_data_system_time_and_sv_tow *)data);
 		break;
 	}
-	case NRF_MODEM_GNSS_AGPS_LOCATION: {
-		print_location((struct nrf_modem_gnss_agps_data_location *)data);
+	case NRF_MODEM_GNSS_AGNSS_LOCATION: {
+		print_location((struct nrf_modem_gnss_agnss_data_location *)data);
 		break;
 	}
 	case NRF_MODEM_GNSS_AGPS_INTEGRITY: {
```
Comment (eivindj-nordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c:175]: `NRF_MODEM_GNSS_AGPS_LOCATION` seem a bit misplaced here now, maybe(?): 
```suggestion
	case NRF_MODEM_GNSS_AGNSS_INTEGRITY_GPS: {
```

File: samples/nrf9160/gnss/src/assistance.c [hunk 3]
```diff
@@ -308,9 +309,9 @@ int assistance_request(struct nrf_modem_gnss_agps_data_frame *agps_request)
 	}
 
 	LOG_INF("Requesting A-GPS data, ephe 0x%08x, alm 0x%08x, flags 0x%02x",
-		agps_request->sv_mask_ephe,
-		agps_request->sv_mask_alm,
-		agps_request->data_flags);
+		(uint32_t)gps_data_need->system[0].sv_mask_ephe,
```
Comment (eivindj-nordic) [inline samples/nrf9160/gnss/src/assistance.c:312]:
```suggestion
		(uint32_t)agnss_request->system[0].sv_mask_ephe,
```

File: include/modem/location.h [hunk 4]
```diff
@@ -222,7 +222,7 @@ struct location_event_data {
 		 * A-GPS notification data frame used by GNSS to let the application know it
 		 * needs new assistance data, used with event LOCATION_EVT_GNSS_ASSISTANCE_REQUEST.
 		 */
-		struct nrf_modem_gnss_agps_data_frame agps_request;
+		struct nrf_modem_gnss_agnss_data_frame agps_request;
```
[the remark this answers is not in the document]
Comment (tokangas) [reply include/modem/location.h:225]: First I started changing all A-GPS related names to A-GNSS, but quickly realized that there's so much changes to be done, that it's better to minimize changes here to just update the GNSS API usage and do rest of the updates later with other PRs. My changes may not be totally consistent across different components, but in my opinion that should not be a big problem either.

File: samples/nrf9160/gnss/src/assistance.c [hunk 5]
```diff
@@ -227,25 +228,25 @@ int assistance_init(struct k_work_q *assistance_work_q)
...
-	if (!agps_request->data_flags) {
+	if (!agnss_request->data_flags) {
 		/* No assistance needed from A-GPS, skip directly to P-GPS. */
 		nrf_cloud_pgps_notify_prediction();
 		return 0;
 	}
 
 	/* P-GPS will handle ephemerides, so skip those. */
-	agps_request->sv_mask_ephe = 0;
+	gps_data_need->system[0].sv_mask_ephe = 0;
```
[the remark this answers is not in the document]
Comment (tokangas) [reply samples/nrf9160/gnss/src/assistance.c:247]: Yes, it should. I did some last minute changes and seems that I didn't compile with all configurations. Some other changes are needed as well.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c (hunk 2 above)
Comment (tokangas) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c:175]: The GNSS API currently contains two types for satellite integrity data, `NRF_MODEM_GNSS_AGPS_INTEGRITY` and `NRF_MODEM_GNSS_AGNSS_INTEGRITY`. `NRF_MODEM_GNSS_AGPS_INTEGRITY` contains data only for GPS satellites while `NRF_MODEM_GNSS_AGNSS_INTEGRITY` can contain data for multiple constellations, but it's only supported from MFW v2.0.0 onwards.

These could be combined in the API, but then the modemlib implementation needs to decide which data type to use in the GNSS RPC API. The decision could be done based on whether integrity is given for only GPS satellites or not, or the implementation could try first using the newer data type and fallback to the older data type if the first request is rejected by the modem.

The application knows from the A-GNSS data request if modem supports assistance for other systems than GPS. It can then request correct assistance data from the network and write integrity assistance using the correct type. This is why I decided to keep two types for integrity data in the GNSS API. That can still be changed, though. What do you think, @eivindj-nordic  and @trantanen?

Comment (eivindj-nordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c:175]: Hmm, good question. 

TL;DR
If there is only one request type (NRF_MODEM_GNSS_AGNSS_INTEGRITY_REQUEST) I think it should be enough with one data type (NRF_MODEM_GNSS_AGNSS_INTEGRITY).

From what I can see, we only have `NRF_MODEM_GNSS_AGNSS_INTEGRITY_REQUEST` that is common for both(?), but we have two types to write. I think it should be ok to try `RPC_GNSS_AGNSS_REQ_DATA_INTEGRITY` first, and if that is not successful try with `RPC_GNSS_AGNSS_REQ_DATA_INTEGRITY_V1`. If we store the result the firs time, we only have to try and fail once. On the other hand I guess this is not a command that is called often, so one extra RPC request now and then should not create much extra load.  

FYI we are looking at the option/need to have separate libmodem binaries for nRF9160 and nRF91x1. Nothing is decided here yet, and timeline is uncertain, but if we choose to do so, then we can select the correct data type dependent on what target it is built for and won't have to try and fail on nRF9160.

**Comment by tokangas** [comment]
I'll continue with this after my summer vacation. I have been working on adding support for QZSS assistance with nRF Cloud, but there are some things which need to be clarified with the cloud side.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c (hunk 2 above)
Comment (tokangas) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c:175]: @eivindj-nordic I tried changing this so, that there would be only one type for integrity data, but realized that this is a bit problematic. The SUPL library gives the data in the old format, so essentially changing the structure of data type 8 (`NRF_MODEM_GNSS_AGPS_INTEGRITY`) causes problems. This would either require a new release of the SUPL library (which is released separately as a binary delivery) or conversion in all applications using the library. Currently, the handler on the application side just passes the data unmodified to `nrf_modem_gnss_agps_write()`:

https://github.com/nrfconnect/sdk-nrf/blob/64ebb31f39903a83a3d611cff1ae332d30c109f1/samples/cellular/modem_shell/src/gnss/gnss.c#L790

But to make it work with the new data structure, the handler would have to convert integrity data from `struct nrf_modem_gnss_agps_data_integrity` to `struct nrf_modem_gnss_agnss_data_integrity` before calling `nrf_modem_gnss_agps_write()`.

I'm leaning towards keeping both types in the API...

Comment (eivindj-nordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c:175]: I see...
I'm not familiar with the SUPL library. Do you know who is maintaining the SUPL library, and how much work it would entail releasing a new version of it? 

It is definitely easier keeping both in that sense. On the other hand I think it would be cleaner to only have one type for integrity data in the API. I don't have a strong opinion of it, perhaps that can be cleaned up in a separate PR later, after we split the binaries for nRF9160 and nRF91x1(?).

Comment (tokangas) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps_utils.c:175]: I am. 😄 We've released a new version when there has been significant fixes or improvements for the library. Because it's released as a separate binary delivery (see https://www.nordicsemi.com/Products/Development-hardware/nRF9160-DK/Download?lang=en#infotabs), there is some extra work involved.

From API usage perspective this is pretty straight forward, I think. `NRF_MODEM_GNSS_AGPS_INTEGRITY` can be used to only write GPS integrity, while `NRF_MODEM_GNSS_AGNSS_INTEGRITY` can be used to write integrity for all supported signals. Normally users don't have to even use these directly, but it's handled by the nRF Cloud library. If you have a look at the implementation in `agps_send_to_modem()` in `nrf_cloud_agps.c`, there's really no extra logic required. GPS integrity is always written using `NRF_MODEM_GNSS_AGPS_INTEGRITY` and it works with both v1.x.x and v2.x.x MFWs. QZSS integrity is always written with `NRF_MODEM_GNSS_AGNSS_INTEGRITY`. It doesn't work with MFW v1.x.x, but it does not support QZSS assistance and does not ask for it either.

I do agree that it would be cleaner to have only one type, but because of the SUPL library dependency I would rather not break backwards compatibility with `NRF_MODEM_GNSS_AGPS_INTEGRITY`.

**Comment by tokangas** [comment]
The PR is now ready for review. Also QZSS assistance support is included.

File: lib/location/method_gnss.c [hunk 6]
```diff
@@ -965,13 +1038,35 @@ static bool method_gnss_agps_expiry_process(const struct nrf_modem_gnss_agps_exp
...
+			/* GNSS supports QZSS PRNs 193...202. Only part of these are currently
+			 * deployed, so even after all available QZSS ephemerides and almanancs
+			 * have been injected, GNSS reports the unused ones as expired.
+			 */
+
+			/* QZSS ephemerides are valid for a maximum of two hours, so no expiry
+			 * is used here.
+			 */
+			if (agps_expiry->sv[i].ephe_expiry == 0) {
+				expired_qzss_ephe_mask |= 1 << (agps_expiry->sv[i].sv_id - 193);
```
Comment (MarkusLassila) [inline lib/location/method_gnss.c:1064]: Should the 193 be defined?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c [hunk 7]
```diff
@@ -54,8 +54,15 @@ bool nrf_cloud_agps_request_in_progress(void)
 	return atomic_get(&request_in_progress) != 0;
 }
 
-int nrf_cloud_agps_request(const struct nrf_modem_gnss_agps_data_frame *request)
+#if IS_ENABLED(CONFIG_NRF_CLOUD_AGPS)
+int nrf_cloud_agps_request(const struct nrf_modem_gnss_agnss_data_frame *request)
```
Comment (MarkusLassila) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:58]: Should this (these?) be renamed as agnss?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c [hunk 8]
```diff
@@ -296,88 +308,146 @@ static int copy_time_and_tow(struct nrf_modem_gnss_agps_data_system_time_and_sv_
...
+			       struct nrf_cloud_apgs_element *src)
+{
+	uint8_t signal_id;
+
+	if ((src == NULL) || (dst == NULL)) {
+		return -EINVAL;
+	}
+
+	signal_id = integrity_signal_get(src->type);
+	if (signal_id == NRF_MODEM_GNSS_SIGNAL_INVALID) {
```
Comment (MarkusLassila) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:347]: Should this zero the possible signal count in dst?

**Review by MarkusLassila — APPROVED** [review]
Looks good from SLM perspective. We will run our GNSS test against the main PR.

@tokangas: Any opinion whether we should add additional AT-command support for the QZSS satellites?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c [hunk 9]
```diff
@@ -296,88 +308,146 @@ static int copy_time_and_tow(struct nrf_modem_gnss_agps_data_system_time_and_sv_
...
+		return 0;
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:348]: `dst` is sent to the modem as all zeroes if the `signal_id` is invalid.
is it required to write the integrity data to the modem if it is invalid/zeroes?

should this function return `-ENODATA` or similar and then `send_to_modem` can be skipped?

**Review by jayteemo — APPROVED** [review]

File: samples/cellular/modem_shell/src/gnss/gnss.c [hunk 10]
```diff
@@ -1509,20 +1551,44 @@ int gnss_set_agps_automatic(bool value)
 #endif
 }
 
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static bool qzss_assistance_is_supported(void)
+{
+	char resp[32];
+
+	if (nrf_modem_at_cmd(resp, sizeof(resp), "AT+CGMM") == 0) {
+		if (strstr(resp, "nRF9161") != NULL) {
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/cellular/modem_shell/src/gnss/gnss.c:1560]: I don't have an answer but I'm thinking if this should be made based on chipset version, or mfw version?

[the remark this answers is not in the document]
Comment (tokangas) [reply samples/cellular/modem_shell/src/gnss/gnss.c:1560]: First, I was going to do this based on the MFW version, but then I realized that it would not work with modem firmware builds which are named differently (for example daily or custom builds). I had similar implementation in https://github.com/nrfconnect/sdk-nrf/pull/12030, there I decided to reverse the check. If the modem identifier starts with nRF9160 we need to use version 2, otherwise the selected version is 3.

[the remark this answers is not in the document]
Comment (trantanen) [reply samples/cellular/modem_shell/src/gnss/gnss.c:1560]: Ok, seems there is no optimal way of doing this.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c (hunk 7 above)
Comment (tokangas) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:58]: That was my idea as well. I was first going to update everything, but it turned out to be too much work to do here. It's better to have follow up PRs later for different components to update rest of the A-GPS references.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c (hunk 8 above)
Comment (tokangas) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:347]: The whole destination structure is zeroed on line 472.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c (hunk 9 above)
Comment (tokangas) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:348]: No, the `send_to_modem()` call could be skipped, but this can actually never happen. `copy_integrity_gnss()` is only called when the data type is `NRF_CLOUD_AGNSS_QZSS_INTEGRITY`. `copy_integrity_gnss()` is supposed to be generic, but it's only used with QZSS currently.

All of the copy functions can also return `-EINVAL` if `src` or `dst` are `NULL`, but the return value is not checked in `agps_send_to_modem()`. We could add the missing checks, but probably `src` and `dst` can not even be `NULL`. I'm starting to think that we should make the copy functions return `void` and maybe add asserts for `src` and `dst`. I think the `signal_id` check in `copy_integrity_gnss()` could also be an assert.

Comment (tokangas) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:348]: Refactored the code a bit. All copy functions now return `void`. I also removed the need for checking the signal ID, because it is always QZSS. I think this is now cleaner. @jayteemo @MarkusLassila

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c (hunk 7 above)
Comment (MarkusLassila) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:58]: Thanks for the clarification. I would resolve this, if I could. :)

**Comment by tokangas** [comment]
> Any opinion whether we should add additional AT-command support to SLM for the QZSS satellites?

@MarkusLassila There are only a few QZSS related options in the API, and those have been present already earlier.

- `nrf_modem_gnss_signal_mask_set()` Can be used to disable QZSS support.
- `nrf_modem_gnss_qzss_nmea_mode_set()` By default, QZSS satellites are not reported in NMEA sentences because the NMEA version we use does not include QZSS. By enabling QZSS NMEA custom mode, QZSS satellites are reported in NMEA using IDs 193...202, but that may be incompatible with some applications. I'm not sure if SLM even uses NMEA, so this may not be relevant.
- `nrf_modem_gnss_qzss_prn_mask_set()` Can be used to set a bitmask to enable only specific QZSS satellites.

I think that's all. Probably there's no need to have AT commands for those in SLM, I think.

**Review by trantanen — APPROVED** [review]
